Allow for custom sharding specs to register their own ops.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #76477
* __->__ #76360

Customized ShardingSpecs could be entirely arbitrary and it would not
be possible to handle ops for those as a result since they might not fit into
the patterns supported by the in-built ShardingSpecs. As a result, we introduce
a framework for a ShardingSpec to override ops as follows:

1) In the dispatch system, if a ShardingSpec has a customized op the registered
op for that ShardingSpec is invoked.
2) As a result, all ChunkShardingSpec specific ops have been moved under that
ShardingSpec.
3) There will be a set of ShardingSpec agnostic ops (ex: elementwise ops) which
will be a set of common ops supported across any ShardingSpec.
4) If an op is not found for a particular ShardingSpec the default set of ops
is searched for that op.

Differential Revision: [D35917912](https://our.internmc.facebook.com/intern/diff/D35917912/)

pritamdamania87 committed:
Allow for custom sharding specs to register their own ops.

Customized ShardingSpecs could be entirely arbitrary and it would not
be possible to handle ops for those as a result since they might not fit into
the patterns supported by the in-built ShardingSpecs. As a result, we introduce
a framework for a ShardingSpec to override ops as follows:

1) In the dispatch system, if a ShardingSpec has a customized op the registered
op for that ShardingSpec is invoked.
2) As a result, all ChunkShardingSpec specific ops have been moved under that
ShardingSpec.
3) There will be a set of ShardingSpec agnostic ops (ex: elementwise ops) which
will be a set of common ops supported across any ShardingSpec.
4) If an op is not found for a particular ShardingSpec the default set of ops
is searched for that op.

Differential Revision: [D35917912](https://our.internmc.facebook.com/intern/diff/D35917912/)

[ghstack-poisoned]

pritamdamania87 committed:
Update on "Allow for custom sharding specs to register their own ops."


Customized ShardingSpecs could be entirely arbitrary and it would not
be possible to handle ops for those as a result since they might not fit into
the patterns supported by the in-built ShardingSpecs. As a result, we introduce
a framework for a ShardingSpec to override ops as follows:

1) In the dispatch system, if a ShardingSpec has a customized op the registered
op for that ShardingSpec is invoked.
2) As a result, all ChunkShardingSpec specific ops have been moved under that
ShardingSpec.
3) There will be a set of ShardingSpec agnostic ops (ex: elementwise ops) which
will be a set of common ops supported across any ShardingSpec.
4) If an op is not found for a particular ShardingSpec the default set of ops
is searched for that op.

Differential Revision: [D35917912](https://our.internmc.facebook.com/intern/diff/D35917912/)

[ghstack-poisoned]

pritamdamania87 committed:
Update on "Allow for custom sharding specs to register their own ops."


Customized ShardingSpecs could be entirely arbitrary and it would not
be possible to handle ops for those as a result since they might not fit into
the patterns supported by the in-built ShardingSpecs. As a result, we introduce
a framework for a ShardingSpec to override ops as follows:

1) In the dispatch system, if a ShardingSpec has a customized op the registered
op for that ShardingSpec is invoked.
2) As a result, all ChunkShardingSpec specific ops have been moved under that
ShardingSpec.
3) There will be a set of ShardingSpec agnostic ops (ex: elementwise ops) which
will be a set of common ops supported across any ShardingSpec.
4) If an op is not found for a particular ShardingSpec the default set of ops
is searched for that op.

Differential Revision: [D35917912](https://our.internmc.facebook.com/intern/diff/D35917912/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -870,6 +885,15 @@ def __truediv__(self, other):
     def __rtruediv__(self, other):
         return handle_torch_function(torch.Tensor.__rdiv__, (self, other), self, other)
 
+    def tanh(self):
```
why we need to special register these three ops in the `ShardedTensor` class?

wanchaol commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec.py:
```diff
@@ -0,0 +1,193 @@
+    get_split_size,
+)
+
+from . import ShardingSpec
```
nit: `from .api import ShardingSpec` to avoid circular imports?

wanchaol commented on torch/distributed/_shard/sharded_tensor/_ops/math_ops.py:
```diff
@@ -4,96 +4,80 @@
+    elif isinstance(rhs, ReplicatedTensor):
+        assert isinstance(lhs, ShardedTensor)
+        st_size = lhs.size()
+        st_meta = lhs.local_shards()[0].metadata
```
(no need to do it in this PR) Do you think for spec agnostic sharded tensor ops, we should consider the case of multiple local shards?

fduwjj replied:
Indeed, some ops have very complicated behavior under the circumstances of multiple shards, which we need to define the expected behavior first. I guess for simple ops like element-wise ops, we already implemented in a spec agnostic way.

wanchaol commented on torch/distributed/_shard/sharding_spec/api.py:
```diff
@@ -92,174 +88,60 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
+        if len(signature(func).parameters) != 3:
+            raise TypeError(
+                f'Custom sharded op function expects signature: '
+                f'(types, args, kwargs), but received '
```
for custom_spec_op, we don't pass process_group, are we assuming that we would use the sharded tensor associated process group for communication in the op implementation? what if two parameters in the module have different process group, should we check that they are in the same pg before doing the computation?

wanchaol commented on torch/distributed/_shard/sharding_spec/api.py:
```diff
@@ -92,174 +88,60 @@ def shard(self, tensor: torch.Tensor, src_rank: int = 0, process_group=None) ->
+
+        class_name = sharding_spec.__qualname__
+        if class_name not in ShardingSpec.custom_ops:
+            ShardingSpec.custom_ops[class_name] = {}
```
why we need a class name as a key here, is it here to store all sharding spec op table? If so, do we have a inheritance rule that if a subclass of a ShardingSpec op can't be find, search it in the parent class instead?

wanchaol commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -39,7 +39,7 @@
 
 # Custom sharded ops
-_SHARDED_OPS: Dict[str, Callable] = {}
+_SHARDED_OPS: Dict[Callable, Callable] = {}
```
should this `_SHRDED_OPS` be removed, and we consolidate the ops into the base `ShardingSpec` instead?

pritamdamania87 replied:
See comment above, it doesn't seem right to have ops registered on an interface class.

wanchaol reviewed (commented):
Looks great! Some questions and suggestions inlined.

pritamdamania87 committed:
Update on "Allow for custom sharding specs to register their own ops."


Customized ShardingSpecs could be entirely arbitrary and it would not
be possible to handle ops for those as a result since they might not fit into
the patterns supported by the in-built ShardingSpecs. As a result, we introduce
a framework for a ShardingSpec to override ops as follows:

1) In the dispatch system, if a ShardingSpec has a customized op the registered
op for that ShardingSpec is invoked.
2) As a result, all ChunkShardingSpec specific ops have been moved under that
ShardingSpec.
3) There will be a set of ShardingSpec agnostic ops (ex: elementwise ops) which
will be a set of common ops supported across any ShardingSpec.
4) If an op is not found for a particular ShardingSpec the default set of ops
is searched for that op.

Differential Revision: [D35917912](https://our.internmc.facebook.com/intern/diff/D35917912/)

[ghstack-poisoned]

pritamdamania87 committed:
Update on "Allow for custom sharding specs to register their own ops."


Customized ShardingSpecs could be entirely arbitrary and it would not
be possible to handle ops for those as a result since they might not fit into
the patterns supported by the in-built ShardingSpecs. As a result, we introduce
a framework for a ShardingSpec to override ops as follows:

1) In the dispatch system, if a ShardingSpec has a customized op the registered
op for that ShardingSpec is invoked.
2) As a result, all ChunkShardingSpec specific ops have been moved under that
ShardingSpec.
3) There will be a set of ShardingSpec agnostic ops (ex: elementwise ops) which
will be a set of common ops supported across any ShardingSpec.
4) If an op is not found for a particular ShardingSpec the default set of ops
is searched for that op.

Differential Revision: [D35917912](https://our.internmc.facebook.com/intern/diff/D35917912/)

[ghstack-poisoned]

fduwjj commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -870,6 +890,15 @@ def __truediv__(self, other):
+    def __getitem__(self, key):
+        return handle_torch_function(torch.Tensor.__getitem__, (self, key), self, key)
+
+    def transpose(self, dim0, dim1):
+        return handle_torch_function(torch.Tensor.transpose, (self, dim0, dim1), self, dim0, dim1)
```
We need meta change for this op right? Btw, transpose has been added in https://github.com/pytorch/pytorch/pull/75374.

fduwjj commented on torch/distributed/_shard/sharded_tensor/api.py:
```diff
@@ -870,6 +890,15 @@ def __truediv__(self, other):
         return handle_torch_function(torch.Tensor.__rdiv__, (self, other), self, other)
 
+    def tanh(self):
+        return handle_torch_function(torch.Tensor.tanh, (self,), self)
```
Nit: can we add this as a new op in the element-wise op?

fduwjj commented on torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/math_ops.py:
```diff
@@ -0,0 +1,69 @@
+                    lhs_spec,
+                    lhs.size(),  # type: ignore[arg-type]
+                    process_group=pg)
+            else:
+                raise RuntimeError("Implicit broadcasting not supported yet!")
```
Nit: we also don't support when sharding dim is different, right?

pritamdamania87 replied:
Broadcasting probably covers the sharding dim mismatch case too since in that case you probably need to do broadcasting across ranks.

fduwjj commented on torch/distributed/_shard/sharded_tensor/_ops/__init__.py:
```diff
@@ -2,7 +2,10 @@
 from .init import kaiming_uniform_, normal_, uniform_, constant_
-from .linear import sharded_linear
+
+# Import all ChunkShardingSpec ops
+from torch.distributed._shard.sharding_spec.chunk_sharding_spec_ops.linear import sharded_linear
+from torch.distributed._shard.sharding_spec.chunk_sharding_spec_ops.embedding import sharded_embedding
+from torch.distributed._shard.sharding_spec.chunk_sharding_spec_ops.embedding_bag import sharded_embedding_bag
+import torch.distributed._shard.sharding_spec.chunk_sharding_spec_ops.math_ops
```
Curious, why not put this into `torch/distributed/_shard/sharding_spec/chunk_sharding_spec_ops/__init__.py` first and we then import that file?

fduwjj reviewed (commented):
Thanks for separating between sharding spec specific ops and general ops. https://github.com/pytorch/pytorch/pull/75374 has been landed and can you kindly rebase on top of it? Thanks!

wanchaol reviewed (approved):
lgtm!

pritamdamania87 committed:
Update on "Allow for custom sharding specs to register their own ops."


Customized ShardingSpecs could be entirely arbitrary and it would not
be possible to handle ops for those as a result since they might not fit into
the patterns supported by the in-built ShardingSpecs. As a result, we introduce
a framework for a ShardingSpec to override ops as follows:

1) In the dispatch system, if a ShardingSpec has a customized op the registered
op for that ShardingSpec is invoked.
2) As a result, all ChunkShardingSpec specific ops have been moved under that
ShardingSpec.
3) There will be a set of ShardingSpec agnostic ops (ex: elementwise ops) which
will be a set of common ops supported across any ShardingSpec.
4) If an op is not found for a particular ShardingSpec the default set of ops
is searched for that op.

Differential Revision: [D35917912](https://our.internmc.facebook.com/intern/diff/D35917912/)

[ghstack-poisoned]

pritamdamania87 committed:
Update on "Allow for custom sharding specs to register their own ops."


Customized ShardingSpecs could be entirely arbitrary and it would not
be possible to handle ops for those as a result since they might not fit into
the patterns supported by the in-built ShardingSpecs. As a result, we introduce
a framework for a ShardingSpec to override ops as follows:

1) In the dispatch system, if a ShardingSpec has a customized op the registered
op for that ShardingSpec is invoked.
2) As a result, all ChunkShardingSpec specific ops have been moved under that
ShardingSpec.
3) There will be a set of ShardingSpec agnostic ops (ex: elementwise ops) which
will be a set of common ops supported across any ShardingSpec.
4) If an op is not found for a particular ShardingSpec the default set of ops
is searched for that op.

Differential Revision: [D35917912](https://our.internmc.facebook.com/intern/diff/D35917912/)

[ghstack-poisoned]

pritamdamania87 commented:
@pytorchbot merge this please

pritamdamania87 committed:
Update on "Allow for custom sharding specs to register their own ops."


Customized ShardingSpecs could be entirely arbitrary and it would not
be possible to handle ops for those as a result since they might not fit into
the patterns supported by the in-built ShardingSpecs. As a result, we introduce
a framework for a ShardingSpec to override ops as follows:

1) In the dispatch system, if a ShardingSpec has a customized op the registered
op for that ShardingSpec is invoked.
2) As a result, all ChunkShardingSpec specific ops have been moved under that
ShardingSpec.
3) There will be a set of ShardingSpec agnostic ops (ex: elementwise ops) which
will be a set of common ops supported across any ShardingSpec.
4) If an op is not found for a particular ShardingSpec the default set of ops
is searched for that op.

Differential Revision: [D35917912](https://our.internmc.facebook.com/intern/diff/D35917912/)

[ghstack-poisoned]

pritamdamania87 commented:
@pytorchbot merge this please
